### ISSUE TITLE
extraVolumeMounts for Spark in IBM Cloud

### ISSUE THREAD

**Issue by mohamedih** [bitnami/charts#2849]
**Which chart**:
Spark in IBM Cloud

**Describe the bug**
How can i add extra volume and extraVolumeMounts to the workers. I tried many combination but it failed. what is the format that i have to use t enter the details?
What this comment mean ?
"Array of extra volume mounts to be added to thehttps://github.com/bitnami/charts/issues/new/choose Spark worker deployment (evaluated as template). Normally used with"

**Comment by andresbono** [comment]
Hi @mohamedih, please check the format with the following example. It will create a volume from a configmap and it will be mounted in the worker pods:

`custom-values.yaml`:
```yaml
## Spark worker specific configuration
##
worker:
  ## Array to add extra volumes
  ##
  extraVolumes:
    - name: extra-volume
      configMap:
        name: extra-configmap
  ## Array to add extra mounts (normally used with extraVolumes)
  ##
  extraVolumeMounts:
    - name: extra-volume
      mountPath: /tmp/test/

```


```
$ kubectl create configmap extra-configmap --from-literal=file='content of the file'
configmap/extra-configmap created

$ helm install spark bitnami/spark -f custom-values.yaml

$ kubectl exec -it spark-worker-0 -- cat /tmp/test/file
content of the file
```

**Comment by mohamedih** [comment]
Thank you @andresbono . In IBM Cloud, I have UI to add my configurations. I do not manually install spark helm chart. I have configuration boxes like below, not sure how can i addextraVolumeMounts and extraVolumes in these configuration box.
<img width="1414" alt="Screen Shot 2020-06-16 at 8 04 44 AM" src="https://user-images.githubusercontent.com/55608955/84772606-2f880a80-afa9-11ea-9a8d-6a6b99a6f128.png">
<img width="1414" alt="Screen Shot 2020-06-16 at 8 04 44 AM" src="https://user-images.githubusercontent.com/55608955/84772614-33b42800-afa9-11ea-8172-64a5cd3b41dd.png">

**Comment by andresbono** [comment]
Oh I see. I haven't tried that UI, but to be safe, I'd use JSON minified format to provide the arrays so they don't contain multiple lines. To continue with the example from my previous comment this would be the format:

```
[ { "name": "extra-volume", "configMap": { "name": "extra-configmap" } } ]
```

```
[ { "name": "extra-volume", "mountPath": "/tmp/test/" } ]
```

**Comment by dkwasny** [comment]
Hello.  I believe I'm hitting the same issue with the `2.4.6` offering.  I tried to replicate the example posted by @andresbono using the same UI as @mohamedih, but hit the following error when I attempt to save the configuration.
![image](https://user-images.githubusercontent.com/1449712/92976792-61e6be00-f459-11ea-966a-d8f4692b2b02.png)

Here are the properties I am attempting to save.
![image](https://user-images.githubusercontent.com/1449712/92976869-922e5c80-f459-11ea-9b71-0d55ae8c3593.png)

Here's the current state of the configmap on my cluster as well.
```
$ kubectl describe configmap extra-configmap
Name:         extra-configmap
Namespace:    default
Labels:       <none>
Annotations:  <none>

Data
====
file:
----
content of the file
Events:  <none>
```

In my initial attempts, I also tried to specify the volume and volume mount using a "flow style" YAML array which seems to work with the helm chart directly, but still fails in the UI.
```
[ { name: extra-volume, mountPath: /tmp/test/ } ]
[ { name: extra-volume, configMap: { name: extra-configmap } } ]
```

I currently can't get any non-empty value to take through the UI right now.  If it comes down to it, I can shift to using the Helm chart directly, but the integrated offering within the IBM cloud would definitely be easier for us.

My only guess is that there's some weird validation going on before the value actually makes it to Helm.

**Comment by dkwasny** [comment]
@andresbono Thank you very much for your quick reply!  Your advice on setting the default on workspace creation allowed me to set the volume fields.

Unfortunately, I'm now hitting an error when I attempt to apply changes to my cluster.
```
 2020/09/14 19:55:45 Terraform apply | Error: error validating "": error validating data: [ValidationError(StatefulSet.spec.template.spec.containers[0].volumeMounts): invalid type for io.k8s.api.core.v1.Container.volumeMounts: got "string", expected "array", ValidationError(StatefulSet.spec.template.spec.volumes): invalid type for io.k8s.api.core.v1.PodSpec.volumes: got "string", expected "array"]
```

Looking at the plan generation log, it looks like the fields I'm setting are indeed being treated as strings.
```
 2020/09/14 19:53:24 Terraform plan |                   extraVolumeMounts: '[ { "name": "extra-volume", "mountPath": "/tmp/kwas-test-1/" } ]'
 2020/09/14 19:53:24 Terraform plan |                   extraVolumes: '[ { "name": "extra-volume", "configMap": { "name": "extra-configmap" } } ]'
```

If I recreate the cluster with my originally intended values as defaults, the mounts seem to be setup correctly.
```
 2020/09/14 20:16:40 Terraform show |                         extraVolumeMounts = [
 2020/09/14 20:16:40 Terraform show |                             {
 2020/09/14 20:16:40 Terraform show |                                 mountPath = "/opt/jars"
 2020/09/14 20:16:40 Terraform show |                                 name      = "jar-volume"
 2020/09/14 20:16:40 Terraform show |                             },
 2020/09/14 20:16:40 Terraform show |                         ]
 2020/09/14 20:16:40 Terraform show |                         extraVolumes      = [
 2020/09/14 20:16:40 Terraform show |                             {
 2020/09/14 20:16:40 Terraform show |                                 name                  = "jar-volume"
 2020/09/14 20:16:40 Terraform show |                                 persistentVolumeClaim = {
 2020/09/14 20:16:40 Terraform show |                                     claimName = "kwas-pvc"
 2020/09/14 20:16:40 Terraform show |                                 }
 2020/09/14 20:16:40 Terraform show |                             },
 2020/09/14 20:16:40 Terraform show |                         ]
```

So I'm good for now...as long as I don't need to change anything 😄

**Comment by andresbono** [comment]
Ok thanks for letting us know, @dkwasny. So the values parameters were treated as arrays by indenting and adding new lines in the fields in JSON format instead of using a minified version?

**Comment by dkwasny** [comment]
@andresbono Not quite.  All I did was set the minified single-line json on workspace creation.  It successfully parsed the list and set it correctly as seen in my earlier comment.  The problem is that any further updates treat the value as a string literal instead of a proper list.

I was informed of the a [Github issue](https://github.ibm.com/blueprint/product/issues/2944) detailing this bug in Blueprint.

Just for my curiosity, I did try updating the value with a multiline json, and got the same error (`got 'string', expected 'array'`).
